### ISSUE TITLE
Datasets and datafiles query request

### ISSUE THREAD

**Issue by maganaluis** [google/ml-metadata#68]
**Proposal**
We would like to use an Artifact Type for Datasets, and another one for Datafiles, then query all DataFiles associated with a particular Dataset. However right now there is no way for me to perform this query. 

We need the ability to query artifacts based on the property values, specifically I'm looking for ```get_artifacts_by_property_value``` where I can pass they key and value I'm looking for. 

The reasoning behind this request is that Datasets can be composed of files, and in restricted projects we might be asked to query and explain the use of each individual file depending on its contents. 

```
class DataFileType(object):
    data_type = metadata_store_pb2.ArtifactType()
    data_type.name = "DataFile_V1.0"
    data_type.properties["data_file_name"] = metadata_store_pb2.STRING
    data_type.properties["data_file_size"] = metadata_store_pb2.STRING
    data_type.properties["data_file_created_by"] = metadata_store_pb2.STRING
    data_type.properties["data_file_created_date"] = metadata_store_pb2.STRING
    data_type.properties["data_file_parent_dataset"] = metadata_store_pb2.STRING
```  

> Note: Another approach would be to allow for artifact hierarchies, though the approach above would solve the problem faster.

**Comment by hughmiao** [comment]
> We need the ability to query artifacts based on the property values, specifically I'm looking for get_artifacts_by_property_value where I can pass they key and value I'm looking for.

thanks for the comments @maganaluis . We are working on a declarative query api to support property queries, which allows to set a predicate in those get_artifacts() calls. 

> Another approach would be to allow for artifact hierarchies, though the approach above would solve the problem faster.

@maganaluis we also have plans to support this too. We plan to release a set of standard base type (e.g., Dataset), where the users' type can declare a sub-type. One can retrieve all subtypes instances giving a parent type identifier. 

Though having this does not solve the problem here, as the query is an instance level listing query.

> query all DataFiles associated with a particular Dataset 

An alternative that is available right now, is to use a ContextType to model these data files groupings, then use the Context APIs, e.g., get_artifacts_by_context, to get those.  For example, say the

An artifact a_0 is the DataSetArtifactType, it has k artifacts {af_i} of DataFileArtifactType,
We can use an ctx_0 of DataBundleContextType, which has attributions to a_0 and {af_i}

**Comment by maganaluis** [comment]
@hughmiao Thank you for you response. I'm glad that this feature is on the roadmap, please let me know if you have an estimate for the release version in which it will be included. 

I'm also aware that the Kubeflow Metadata group is planning to add support for Project, Product, Workflow and Run so I'll keep an eye on that front as well. 

Also thank you for the workaround.

**Comment by hughmiao** [comment]
@maganaluis sg. thanks for updating the issue. ETA EoQ4. Let's keep the issue open before that.

**Comment by redramen** [comment]
Hey @hughmiao wondering if the EoQ4 ETA is still accurate? We're building atop MLMD at Twitter and are using contexts as a work around to querying. While this works for our initial use case, this is not ideal .

**Comment by hughmiao** [comment]
@redramen we are close to surface pagination with limit and order by in get_contexts/artifacts/executions calls, hopefully to include it in the next 0.26 release. EoQ4 for oss full filtering predicate in python client will be difficult. Please feel free to let us know your main use cases and timelines, which will help to prioritize the related work. 

/cc @dushyanthsc

**Comment by redramen** [comment]
Our main use case is we want to query executions or artifacts using properties defined ahead of time. We could get around this by doing filtering client side with the python client or via making the sql queries ourselves. Are there any updated ETAs or rough estimates for when a query API could be released?

**Comment by hughmiao** [comment]
> Our main use case is we want to query executions or artifacts using properties defined ahead of time.  We could get around this by doing filtering client side with the python client or via making the sql queries ourselves. 

@redramen thanks for letting us know the use case. We will improve the get_artifacts/executions/contexts with list options (e.g., limit, order by creation-time/update-time/id) in the coming release. hope these can help writing client side filtering.

> An alternative we're also using is setting up contexts which associate/attribute the correct executions/artifacts but it isn't the best approach since we would need to set up these contexts ahead of time.
> E.g. say a query to get all the executions in a running state or something along those lines.

True, using context solves part of the problem. Note we view scoping subgraphs with contexts a best practice to track  and query metadata generated by many pipelines.

> Are there any updated ETAs or rough estimates for when a query API could be released?

ETA will be extended to Q1. If you look at the recent commits, a filter predicate will be added to the list options to add conditions with SQL syntax. We are working on oss this which is blocked by the tf / zetasql dependency changes in the c++ layer. please stay tuned.

**Comment by hughmiao** [comment]
@arllanos yes, it is not included in 1.0 yet (fwiw we sync the version number with tfx). About the progress, we are working on it actively this Q, recently we are dropping tf workspace dependency in cc to unblock the query lib oss. ETA Jul.

/cc @BrianSong
